reference/3.0/Microsoft.PowerShell.Core/About/about_Ref.md

```diff
@@ -5,72 +5,156 @@ locale:  en-us
-returns to the calling method.
+## Long description
 
-To use referencing, the parameter must be a reference variable. If it is not,
-an InvalidArgument exception is thrown.
+You can pass variables to methods *by reference* or *by value*.
 
-The parameters used in method invocations must match the type required by the
-methods.
+In the following example, a method attempts to change the value of a
+variable that is passed into it, and fails.
 
-Examples:
+When you pass a variable *by value*, you are passing a copy of the data.
 
 ```powershell
-PS> function swap([ref]$a,[ref]$b)
->> {
->>     $a.value,$b.value = $b.value,$a.value
->> }
-
-PS> $a = 1
-PS> $b = 10
-PS> $a,$b
-1
-10
-PS> swap ([ref]$a) ([ref]$b)
-PS> $a,$b
+Function Test($data)
+{
+    $data = 3
+}
+
+$var = 10
+Test -data $var
+$var
+```
+
+```output
 10
-1
+```
+
+In the following example, a variable containing a `Hashtable` is passed to a
+method. `Hashtable` is an object type so by default it is passed to the
+method *by reference*.
+
+When passing a variable *by reference*, the method can change the data and
+that change will persist after the method executes.
+
+```powershell
+Function Test($data)
+{
+    $data.Test = "New Text"
+}
+
+$var = @{}
+Test -data $var
+$var
+```
+
+```output
+Name                           Value
+----                           -----
+Test                           New Text
+```
+
+The method adds its own key, which you can still view after the
+method executes.
+
+### Writing methods to accept reference parameters
+
+You can code your methods to take a parameter as a reference, regardless of
+the type of data passed. This requires that you specify the parameters type
+as `System.Management.Automation.PSReference`, or `[ref]`.
+
+When using references, you will need to use the `Value` property of the
+`System.Management.Automation.PSReference` type to access your data.
+
+```powershell
+Function Test([ref]$data)
+{
+    $data.Value = 3
+}
+```
+
+To pass a variable to a parameter that expects a reference, you must type
+cast your variable as a reference.
 
-PS C:\ps-test> function double
->> {
->>     param ([ref]$x) $x.value = $x.value * 2
->> }
+> [!NOTE]
+> The brackets and parenthesis are BOTH required.
 
-PS> $number = 8
+```powershell
+$var = 10
+Test -data ([ref]$var)
+$var
+```
+
+```output
+3
+```
+
+### Passing references to .NET methods
+
+Certain .NET methods may require you to pass a variable as a reference. When
+the method's definition uses the keywords `in`, `out`, or `ref` on a
+parameter, it expects a reference.
+
+```powershell
+[int] | Get-Member -Static -Name TryParse
+```
+
+```output
+Name     MemberType Definition
+----     ---------- ----------
+TryParse Method     static bool TryParse(string s, [ref] int result)
+```
+
+For the `TryParse` method, you can attempt to parse a string as an integer.
+If the method succeeds, the method will return true, and the result will
```

Comment:
If the method succeeds, it returns `$true` and the result is stored in the variable that you passed **by reference**.

Reply:
check